### ISSUE TITLE
guestbook e2e appears to strand resources sometimes

### ISSUE THREAD

**Issue by brendandburns** [GoogleCloudPlatform/kubernetes#4961]
```
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:47
Error running &{/jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/guestbook.sh [/jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/guestbook.sh] []  <nil> Project: kubernetes-jenkins
Zone: us-central1-f
WARNING: this test is a no op that only attempts to launch guestbook resources.
current-context: "kubernetes-jenkins_e2e-test-jenkins"
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
frontend-controller
redis-master-controller
redis-master
redis-slave-controller
redisslave
F0302 21:11:04.205794   16437 create.go:83] Timeout: request did not complete within allowed duration
 Project: kubernetes-jenkins
Zone: us-central1-f
WARNING: this test is a no op that only attempts to launch guestbook resources.
current-context: "kubernetes-jenkins_e2e-test-jenkins"
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
frontend-controller
redis-master-controller
redis-master
redis-slave-controller
redisslave
F0302 21:11:04.205794   16437 create.go:83] Timeout: request did not complete within allowed duration
 [] <nil> 0xc208201120 exit status 255 <nil> true [0xc2081100f8 0xc208110120 0xc208110120] [0xc2081100f8 0xc208110120] [0xc208110118] [0x5d1850] 0xc20822cea0}:
Command output:
Project: kubernetes-jenkins
Zone: us-central1-f
WARNING: this test is a no op that only attempts to launch guestbook resources.
current-context: "kubernetes-jenkins_e2e-test-jenkins"
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
frontend-controller
redis-master-controller
redis-master
redis-slave-controller
redisslave
F0302 21:11:04.205794   16437 create.go:83] Timeout: request did not complete within allowed duration
```

Subsequent runs are failing due to the target pool being stranded.

**Comment by ixdy** [comment]
Was just going to open an issue about this. Log from subsequent runs:
```
 Error running &{/jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/guestbook.sh [/jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/guestbook.sh] []  <nil> Project: kubernetes-jenkins
  Zone: us-central1-f
  WARNING: this test is a no op that only attempts to launch guestbook resources.
  current-context: "kubernetes-jenkins_e2e-test-jenkins"
  Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
  frontend-controller
  redis-master-controller
  redis-master
  redis-slave-controller
  redisslave
  F0302 23:27:26.341903   19201 create.go:83] The resource 'projects/kubernetes-jenkins/regions/us-central1/targetPools/e2e-test-jenkins-frontend' already exists
   Project: kubernetes-jenkins
  Zone: us-central1-f
  WARNING: this test is a no op that only attempts to launch guestbook resources.
  current-context: "kubernetes-jenkins_e2e-test-jenkins"
  Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
  frontend-controller
  redis-master-controller
  redis-master
  redis-slave-controller
  redisslave
  F0302 23:27:26.341903   19201 create.go:83] The resource 'projects/kubernetes-jenkins/regions/us-central1/targetPools/e2e-test-jenkins-frontend' already exists
   [] <nil> 0xc20818e6a0 exit status 255 <nil> true [0xc208108058 0xc208108080 0xc208108080] [0xc208108058 0xc208108080] [0xc208108078] [0x5d1850] 0xc2081ca240}:
  Command output:
  Project: kubernetes-jenkins
  Zone: us-central1-f
  WARNING: this test is a no op that only attempts to launch guestbook resources.
  current-context: "kubernetes-jenkins_e2e-test-jenkins"
  Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
  frontend-controller
  redis-master-controller
  redis-master
  redis-slave-controller
  redisslave
  F0302 23:27:26.341903   19201 create.go:83] The resource 'projects/kubernetes-jenkins/regions/us-central1/targetPools/e2e-test-jenkins-frontend' already exists
```

**Comment by roberthbailey** [comment]
Since it didn't appear that anyone else had done so yet, I manually whacked that target pool to try and get the e2e tests passing again.

**Comment by jlowdermilk** [comment]
This is happening the trap for cleanup is getting skipped somehow; guestbook creates an external load balancer for the frontend service, which isn't torn down with the cluster (unless service is explicitly deleted). Porting this to a ginkgo test should help with that.

**Comment by ixdy** [comment]
This failed again today in the same way.

```
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:47
Error running &{/jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/guestbook.sh [/jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/guestbook.sh] []  <nil> Project: kubernetes-jenkins
Zone: us-central1-f
WARNING: this test is a no op that only attempts to launch guestbook resources.
current-context: "kubernetes-jenkins_e2e-test-jenkins"
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
frontend-controller
redis-master-controller
redis-master
redis-slave-controller
redisslave
F0303 20:07:38.721200   13658 create.go:83] Timeout: request did not complete within allowed duration
Project: kubernetes-jenkins
Zone: us-central1-f
WARNING: this test is a no op that only attempts to launch guestbook resources.
current-context: "kubernetes-jenkins_e2e-test-jenkins"
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
frontend-controller
redis-master-controller
redis-master
redis-slave-controller
redisslave
F0303 20:07:38.721200   13658 create.go:83] Timeout: request did not complete within allowed duration
[] <nil> 0xc20801f120 exit status 255 <nil> true [0xc20813e3b0 0xc20813e3d0 0xc20813e3d0] [0xc20813e3b0 0xc20813e3d0] [0xc20813e3c8] [0x5d1fc0] 0xc208144ba0}:
Command output:
Project: kubernetes-jenkins
Zone: us-central1-f
WARNING: this test is a no op that only attempts to launch guestbook resources.
current-context: "kubernetes-jenkins_e2e-test-jenkins"
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/hack/e2e-suite/../../cluster/gce/../../examples/guestbook
frontend-controller
redis-master-controller
redis-master
redis-slave-controller
redisslave
F0303 20:07:38.721200   13658 create.go:83] Timeout: request did not complete within allowed duration
```

Did something change recently that might be causing these timeouts?

**Comment by jlowdermilk** [comment]
@ixdy, those timeout failures are a distinct issue from the target pool leakage.

**Comment by piosz** [comment]
cc @jszczepkowski who is working on #4630

**Comment by piosz** [comment]
Merge of my PR #4588 should help. In the current version cleanup isn't properly called in case of failure. I added a function "teardown" which is called on every exit of the test - no matter if succeed or failed.

**Comment by ixdy** [comment]
This is still leaking target pools on Jenkins occasionally.

**Comment by jlowdermilk** [comment]
#5045 should help make cleanup consistent in the event of test failures.

**Comment by piosz** [comment]
#5177 fixed it.

**Comment by piosz** [comment]
Fixed by #5177, #5196 and #5262
